docs/architecture/adr-012-ibc-relayer-v2.md

```diff
@@ -0,0 +1,829 @@
+send them together using
+`SendIbcMessagesWithUpdateClient<SendIbcMessagesToChain>`. Through this, the
+batched version of the message sender can save cost and performance on sending
+IBC messages, because the relayer would only attach one UpdateClient message
+alongside multiple messages that are batched together.
+
+With the declarative nature of context-generic programming, users would be able
+to easily customize on different strategies of sending IBC messages,
+as well as building update client messages. CGP also helps propagate the
+additional constraints of components to the concrete context implementer.
+For instance, if
+[`SendMessagesToBatchWorker`](ibc_relayer_framework::full::batch::message_sender::SendMessagesToBatchWorker)
+is used, the relay context is
+required to provide MPSC channels that can be used for sending messages to the
+batch worker. On the other hand, if only `MinimalIbcMessageSender` is used,
+the relay context can remove the burden of having to provide an MPSC channel.
+
+### Async Concurrency
+
+The v1 relayer was developed at a time when Rust's async/await infrastructure
+was not yet ready for use. As a result, the v1 relayer uses thread-based
+concurrency, by spawning limited number of threads and manually multiplexing
+multiple tasks in each thread.
+
+As time moves forward, the async/await feature in Rust has become mature enough,
+and the new relayer is able to make use of async tasks to manage the concurrency
+for relaying packets. At its core, the relaying of packets is done through the
+following interface:
+
+```rust,ignore
+pub trait PacketRelayer<Relay>: Async
+where
+    Relay: HasRelayTypes,
+{
+    async fn relay_packet(relay: &Relay, packet: &Relay::Packet) ->
+        Result<(), Relay::Error>;
+}
+```
+
+The
+[`PacketRelayer`](ibc_relayer_framework::base::relay::traits::packet_relayer::PacketRelayer)
+trait allows the handling of a single IBC packet at a time.
+When multiple IBC packets need to be relayed at the same time, the relayer
+framework allows multiple async tasks to be spawned at the same time, with each
+task sharing the same relay context but doing the relaying for different packets.
+
+To optimize for efficiency, the relay context can switch the strategy for
+batching messages and transactions at the lower layers without affecting the
+logic for the packet relayers themselves. From the perspective of the packet
+relayer implementation, the relay context appears to be exclusively owned by the
+packet relayer, and it is not aware of other concurrent tasks running.
+
+The relayer framework uses multiple layers of optimizations to improve the
+efficiency of relaying IBC packets. The first layer performs message
+batching per relay context, by collecting messages being sent over a relay
+context within a time frame and sending them all as a single batch if it does not
+exceed the batch size limit. This layer does the batching per relay context,
```

Comment:
I am not sure to understand what the relayer framework and the optimisation layers referred here are. At the end it is specified that these optimisation layers are optional and new optimisations can easily be added. Does this mean that the three optimisation layers are examples of what can be done ?

Reply:
Yes, the optimizations listed are the default that is provided, but advance users can opt out of the default if they have other options. I have clarified the given paragraph.